lib/internal/streams/compose.js

```diff
@@ -94,63 +121,119 @@ module.exports = function compose(...streams) {
+      d._write = function(chunk, encoding, callback) {
+        if (head.write(chunk, encoding)) {
+          callback();
+        } else {
+          ondrain = callback;
+        }
+      };
 
-    tail.on('finish', function() {
-      if (onfinish) {
-        const cb = onfinish;
-        onfinish = null;
-        cb();
-      }
-    });
+      d._final = function(callback) {
+        head.end();
+        onfinish = callback;
+      };
+
+      head.on('drain', function() {
+        if (ondrain) {
+          const cb = ondrain;
+          ondrain = null;
+          cb();
+        }
+      });
+    } else if (isWebStream(head)) {
+      const writable = isTransformStream(head) ? head.writable : head;
+      const writer = writable.getWriter();
+
+      d._write = async function(chunk, encoding, callback) {
+        try {
+          await writer.ready;
+          writer.write(chunk).catch(() => {});
+          callback();
+        } catch (err) {
+          callback(err);
+        }
+      };
+
+      d._final = async function(callback) {
+        try {
+          await writer.ready;
+          writer.close().catch(() => {});
+          onfinish = callback;
+        } catch (err) {
+          callback(err);
+        }
+      };
+    }
+    if (isNodeStream(tail)) {
+      tail.on('finish', function() {
+        if (onfinish) {
+          const cb = onfinish;
+          onfinish = null;
+          cb();
+        }
+      });
+    } else if (isWebStream(tail)) {
+      const readable = isTransformStream(tail) ? tail.readable : tail;
+      eos(readable, () => {
+        if (onfinish) {
+          const cb = onfinish;
+          onfinish = null;
+          cb();
+        }
+      });
+    }
   }
 
   if (readable) {
-    tail.on('readable', function() {
-      if (onreadable) {
-        const cb = onreadable;
-        onreadable = null;
-        cb();
-      }
-    });
+    if (isNodeStream(tail)) {
+      tail.on('readable', function() {
+        if (onreadable) {
+          const cb = onreadable;
+          onreadable = null;
+          cb();
+        }
+      });
 
-    tail.on('end', function() {
-      d.push(null);
-    });
+      tail.on('end', function() {
+        d.push(null);
+      });
 
-    d._read = function() {
-      while (true) {
-        const buf = tail.read();
+      d._read = function() {
+        while (true) {
+          const buf = tail.read();
+          if (buf === null) {
+            onreadable = d._read;
+            return;
+          }
 
-        if (buf === null) {
-          onreadable = d._read;
-          return;
+          if (!d.push(buf)) {
+            return;
+          }
         }
+      };
+    } else if (isWebStream(tail)) {
+      const readable = isTransformStream(tail) ? tail.readable : tail;
+      const reader = readable.getReader();
+      d._read = async function() {
+        while (true) {
+          try {
+            const { value, done } = await reader.read();
+            if (done) {
```

Comment:
Technically this doesn't guarantee that if done is true no value is returned but I'm fine with this until someone complains

Reply:
Oh 😅
I think then could rewrite it as 
`d.push(value)
d.push(null)` ?

Reply:
Have updated the flow here, should be better now